src/builder/ruleset_builder.cpp

```diff
@@ -0,0 +1,198 @@
+        for (const auto &[id, spec] : global_config.user_rules) {
+            rule_builder builder{id, spec};
+            if (builder.is_enabled()) {
+                final_user_rules_.emplace(builder.build(*actions_));
+            }
+        }
+    }
+
+    // Generate exclusion filters targetting all final rules
+    if ((current_changes & filters_update) != change_set::none) {
+        rule_filters_.clear();
+        input_filters_.clear();
+
+        // First generate rule filters
+        for (const auto &[id, filter] : global_config.rule_filters) {
+            auto rule_targets = resolve_references(filter.targets, final_base_rules_);
+            rule_targets.merge(resolve_references(filter.targets, final_user_rules_));
+
+            auto filter_ptr = std::make_shared<exclusion::rule_filter>(
+                id, filter.expr, std::move(rule_targets), filter.on_match, filter.custom_action);
+            rule_filters_.emplace(id, filter_ptr);
+        }
+
+        // Finally input filters
+        for (const auto &[id, filter] : global_config.input_filters) {
+            auto rule_targets = resolve_references(filter.targets, final_base_rules_);
+            rule_targets.merge(resolve_references(filter.targets, final_user_rules_));
+
+            auto filter_ptr = std::make_shared<exclusion::input_filter>(
+                id, filter.expr, std::move(rule_targets), filter.filter);
+            input_filters_.emplace(id, filter_ptr);
+        }
+    }
+
+    // Generate new processors
+    if ((current_changes & processors_update) != change_set::none) {
+        preprocessors_.clear();
+        postprocessors_.clear();
+
+        for (const auto &[id, spec] : global_config.processors) {
+            auto proc = processor_builder::build(id, spec, global_config.scanners);
+            if (spec.evaluate) {
+                preprocessors_.emplace(proc->get_id(), std::move(proc));
+            } else {
+                postprocessors_.emplace(proc->get_id(), std::move(proc));
+            }
+        }
+    }
+
+    if ((current_changes & change_set::rule_data) != change_set::none) {
+        rule_matchers_.clear();
+        for (const auto &[id, spec] : global_config.rule_data) {
+            rule_matchers_.emplace(id, matcher_builder::build(spec));
+        }
+    }
+
+    if ((current_changes & change_set::exclusion_data) != change_set::none) {
+        exclusion_matchers_.clear();
+        for (const auto &[id, spec] : global_config.exclusion_data) {
+            exclusion_matchers_.emplace(id, matcher_builder::build(spec));
+        }
+    }
+
+    auto rs = std::make_shared<ruleset>();
+    rs->insert_rules(final_base_rules_.items(), final_user_rules_.items());
+    rs->insert_filters(rule_filters_);
+    rs->insert_filters(input_filters_);
+    rs->insert_preprocessors(preprocessors_);
+    rs->insert_postprocessors(postprocessors_);
+    rs->rule_matchers = rule_matchers_;
+    rs->exclusion_matchers = exclusion_matchers_;
```

Comment:
same comment as with the rules. These are always build all together, so they could all be behind one single (for instance) `shared_ptr<const map<string, unique_ptr<matcher::base>>>`.

It seems the only case where this doesn't apply is the `scanners`, where a copy of the vector is needed because `global_config.scanners` can be modified at any point and `actions`, which already do it that way

Reply:
Same as above, I'll address this in the next PR.